### DISCUSSION TITLE
EFDT debug_one sometimes doesn't work

### DISCUSSION THREAD

**Discussion by FedericoMz** [online-ml/river#1247] [Q&A]
EDIT: I should have posted this as an issue, I apologize. Reposted [here](https://github.com/online-ml/river/issues/1248)
---
Hello! I've been experimenting with Extremely Fast Decision Tree. 

I have a stream of records and labels. After training my model on each record and label, I use the _debug_one_ method on the previously-labeled records to see if the decision rule had changed. Unfortunately, there is always a problematic record which gives my this error.

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[32], line 1
----> 1 model.debug_one(x_past)

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/hoeffding_tree.py:364, in HoeffdingTree.debug_one(self, x)
    361         except KeyError:
    362             child_index, _ = node.most_common_path()
--> 364         _print(node.repr_branch(child_index))
    366 return buffer.getvalue()

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/nodes/branch.py:195, in NominalMultiwayBranch.repr_branch(self, index, shorten)
    194 def repr_branch(self, index: int, shorten=False):
--> 195     feat_val = self._r_mapping[index]
    197     if shorten:
    198         return str(feat_val)

KeyError: 8
```

It's worth noting that both _predict_one_ and _predict_proba_one_ work. However, the issue seems to affect _draw_ too.

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[26], line 1
----> 1 model.draw()

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/hoeffding_tree.py:477, in HoeffdingTree.draw(self, max_depth)
    471     dot.node(f"{child_no}", text, fillcolor=fillcolor, style="filled")
    473     if parent_no is not None:
    474         dot.edge(
    475             f"{parent_no}",
    476             f"{child_no}",
--> 477             xlabel=parent.repr_branch(branch_index, shorten=True),
    478         )
    480 return dot

File ~/miniforge3/envs/fawos/lib/python3.11/site-packages/river/tree/nodes/branch.py:195, in NominalMultiwayBranch.repr_branch(self, index, shorten)
    194 def repr_branch(self, index: int, shorten=False):
--> 195     feat_val = self._r_mapping[index]
    197     if shorten:
    198         return str(feat_val)

KeyError: 6
```

If I train the model a second time on the problematic record, I can use _debug_one_, but the issue persists with _draw_.

I don't think the problem lies with the data, because:
1) The model learns from the problematic record, and _debug_one_ works as aspected. The problem arises after training the model on some new records;
2) I tried multiple set of data and there's always a (different) problematic record.

**Comment by MaxHalford** [comment]
There's no need to open a discussion an issue and a discussion. I'll close this discussion, and we'll answer in the issue.
